compiler/rustc_middle/src/middle/effective_visibility.rs

```diff
@@ -0,0 +1,133 @@
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def_id::LocalDefId;
+use rustc_middle::middle::privacy::AccessLevel;
+use rustc_data_structures::stable_hasher::{HashStable, StableHasher};
+use rustc_macros::HashStable;
+use rustc_query_system::ich::StableHashingContext;
+use crate::ty::Visibility;
+use crate::ty::DefIdTree;
+
+#[derive(Debug, Clone)]
+pub struct EffectiveVisibilities {
+    pub map: FxHashMap<LocalDefId, EffectiveVisibilityValue>,
+}
+
+#[derive(Clone, Debug, Copy, HashStable)]
+pub struct EffectiveVisibilityValue {
+    vis: [Option<LocalDefId>; 4]
```

Comment:
It's better to use fields with human-readable names instead of an array here.

If you need to loop over fields in `fn min` you can still do it with `for something in [&self.public, &self.exported, &self.reachable, &self.reachable_impl_trait]`.